File: LeetCode/[M] - 1268. Search Suggestions System/solution.py
from typing import List


class Node:
    def __init__(self):
        self.words = []
        self.childrens = {}


class Solution:
    def suggestedProducts(
        self, products: List[str], searchWord: str
    ) -> List[List[str]]:
        products.sort()
        root = Node()
        node = root
        for product in products:
            for char in product:
                if char not in node.childrens:
                    node.childrens[char] = Node()
                node = node.childrens[char]
                node.words.append(product)
            node = root

        res = [[] for _ in range(len(searchWord))]
        for i in range(len(searchWord)):
            if searchWord[i] not in node.childrens:
                return res
            node = node.childrens[searchWord[i]]
            res[i] = node.words[:3]

        return res

File: LeetCode/[M] - 1268. Search Suggestions System/solution.go
package main

import "slices"

func suggestedProducts(products []string, searchWord string) [][]string {
	slices.Sort(products)

	searchSystem := make(map[string][]string)

	for _, product := range products {
		for i := range product {
			key := product[:i+1]
			searchSystem[key] = append(searchSystem[key], product)
		}
	}

	res := make([][]string, len(searchWord))
	for i := range searchWord {
		key := searchWord[:i+1]
		if _, ok := searchSystem[key]; !ok {
			return res
		}
		if len(searchSystem[key]) > 3 {
			searchSystem[key] = searchSystem[key][:3]
		}
		res[i] = searchSystem[key]
	}
	return res
}
